# Load a pretrained decoder and keep training it

In [notebook 05](05-train-a-language-model.ipynb) we trained a language model from random weights. But most of the time, you would start from a model someone else already trained, on far more data than we have. So in this notebook, we load SmolLM2-135M into Dew's own `CausalTransformer`. It is a small Llama-style model from Hugging Face, described in the paper ["SmolLM2: When Smol Goes Big" by Allal et al. (2025)](https://arxiv.org/abs/2502.02737). We continue its training on Shakespeare for a few hundred steps and export the result back to the Hugging Face format. At the end, we load the export with the `transformers` library and check that it predicts the same next token as Dew.

**The notebook expects one NVIDIA GPU and takes about three minutes on a Colab L4. The first run downloads the checkpoint (about 270 MB). On a GPU it needs about 8.5 GB of host memory, because JAX and Orbax stage about 2 GB through host memory the first time they save a checkpoint.**

## Install Dependencies

`interop` adds safetensors support, and `torch` is only for the last cell, where `transformers` reads the exported checkpoint.

In [ ]:
%pip install -q "dewml[cuda13,interop,streaming] @ git+https://github.com/AshishKumar4/dew" torch

## Settings

Note the learning rate, 3e-5, a lot smaller than the 1e-3 of notebook 05; the continued training section explains why.

In [1]:
import os

CHECKPOINT = "HuggingFaceTB/SmolLM2-135M"
SEQUENCE_LENGTH = 256
BATCH_SIZE = 8
STEPS = 300
LEARNING_RATE = 3e-5
MAX_NEW_TOKENS = 60
PROMPT = "ROMEO:"
RUN_DIR = "runs/08-continued"
EXPORT_DIR = "runs/08-exported"
SEED = 0

# The tutorial test sets DEW_TUTORIAL_SMOKE=1 to run every cell in minutes on a CPU.
if os.environ.get("DEW_TUTORIAL_SMOKE") == "1":
    SEQUENCE_LENGTH, BATCH_SIZE, STEPS, MAX_NEW_TOKENS = 64, 2, 2, 8


In [2]:
import os

import jax
import jax.numpy as jnp
import numpy as np

print(jax.devices())

[CudaDevice(id=0)]


## Loading the checkpoint

`PretrainedDecoder.load` takes a Hub repo id or a local folder in the Hugging Face layout. It translates the checkpoint's `config.json` into `CausalTransformer` settings and its weights into Dew's parameter tree. The weights stay in float32, and the model computes in the `dtype` we ask for. **If the config has a setting that changes the computation and Dew has no equivalent for it, `load` raises an error naming it, instead of silently dropping it.**

It returns a `PretrainedDecoder`, the kind of `Pretrained` a decoder checkpoint loads as, with the model on `.model`, its weights on `.variables`, and the Dew settings it derived on `.model_config`.

In [3]:
from dew.interop import PretrainedDecoder

pretrained = PretrainedDecoder.load(CHECKPOINT, dtype=jnp.bfloat16, attention_impl="auto", max_seq_len=512)
model, variables = pretrained.model, pretrained.variables
n_params = sum(x.size for x in jax.tree_util.tree_leaves(variables))
print(f"{n_params / 1e6:.0f}M parameters")
print({key: pretrained.model_config[key] for key in
       ("vocab_size", "emb_features", "num_layers", "num_heads", "num_kv_heads", "head_dim")})

19:45:46 WARNING dew.telemetry.devices: Dew was imported after the JAX backend opened; its numerical policy cannot take effect. Restart with XLA_FLAGS=--xla_allow_excess_precision=false set before importing JAX.


/content/tutorials-main6329435-recovery/job-1/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Fetching 7 files:   0%|          | 0/7 [00:00<?, ?it/s]

Fetching 7 files:  14%|█▍        | 1/7 [00:00<00:01,  3.77it/s]

Fetching 7 files: 100%|██████████| 7/7 [00:00<00:00, 23.15it/s]

Fetching 7 files:   0%|          | 0/7 [00:00<?, ?it/s]

Fetching 7 files: 100%|██████████| 7/7 [00:00<00:00, 1629.04it/s]

[dry-run] Fetching 10 files:   0%|          | 0/10 [00:00<?, ?it/s]

[dry-run] Fetching 10 files:  90%|█████████ | 9/10 [00:00<00:00, 36.83it/s]

[dry-run] Fetching 10 files: 100%|██████████| 10/10 [00:00<00:00, 39.53it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 1 files: 100%|██████████| 1/1 [00:03<00:00,  3.69s/it]

Fetching 1 files: 100%|██████████| 1/1 [00:03<00:00,  3.69s/it]

135M parameters
{'vocab_size': 49152, 'emb_features': 576, 'num_layers': 30, 'num_heads': 9, 'num_kv_heads': 3, 'head_dim': 64}


The model was trained on the ids of its own tokenizer, so we must use that tokenizer too; the byte tokenizer of notebook 05 would feed it ids that mean something else entirely in its vocabulary. `HFTokenizer` wraps any Hugging Face tokenizer behind the same `encode` and `decode` as the byte tokenizer.

In [4]:
from dew.data import HFTokenizer

tokenizer = HFTokenizer(CHECKPOINT)
print("vocabulary:", tokenizer.vocab_size)
print(tokenizer.encode(PROMPT))

vocabulary: 49152
[3911, 3945, 63, 42]


## Generating before training

Before we train anything, let's see what the model writes as it is. SmolLM2 learned from web text, code and textbooks, so it knows what a play looks like, but it does not write like Shakespeare. We generate greedily, taking the most likely token every time, so the output is the same on every run.

In [5]:
from dew.sampling import Sampling, generate

prompt = jnp.asarray([tokenizer.encode(PROMPT)], jnp.int32)
before = generate(model, variables, prompt, max_new_tokens=MAX_NEW_TOKENS,
                  key=jax.random.key(0), sampling=Sampling(temperature=0.0))
print(tokenizer.decode(before.tokens[0]))

ROMEO: I think that's a good point.

JOHNSON: I think that's a good point.

MARTIN: I think that's a good point.

JOHNSON: I think that's a good point.

MARTIN: I think


## Data Pipeline

`dew.data.load` reads and caches Tiny Shakespeare as in notebook 05, but with SmolLM2's tokenizer this time. Passing `tokenizer=CHECKPOINT` uses the same 49,152-entry vocabulary as the loaded model.

In [6]:
import dew.data

data = dew.data.load(
    "hf/winglian/tiny-shakespeare", batch=BATCH_SIZE, tokenizer=CHECKPOINT, seq_len=SEQUENCE_LENGTH)

Generating train split:   0%|          | 0/472 [00:00<?, ? examples/s]

Generating train split: 100%|██████████| 472/472 [00:00<00:00, 134072.29 examples/s]

Generating test split:   0%|          | 0/49 [00:00<?, ? examples/s]

Generating test split: 100%|██████████| 49/49 [00:00<00:00, 33987.25 examples/s]

In [7]:
print("training windows:", data.records)

training windows: 1441


## Continued training

`LMObjective` takes the loaded weights through `variables=`, so training starts from them instead of from a fresh random draw. It keeps no moving average of the weights, so validation scores the weights that trained. The learning rate is small so that the model adapts to Shakespeare without forgetting what it already knew. The validation perplexity at steps 100, 200 and 300 shows the adaptation.

In [8]:
from dew import Checkpoints, Trainer
from dew.config import OptimConfig
from dew.objectives.lm import LMObjective, Perplexity

objective = LMObjective(model, SEQUENCE_LENGTH, variables=variables)
trainer = Trainer(objective, OptimConfig(learning_rate=LEARNING_RATE), key=jax.random.key(SEED),
                  checkpoints=Checkpoints(RUN_DIR))
state = trainer.fit(data, steps=STEPS, log_every=50, eval_every=100, checkpoint_every=STEPS,
                    metrics=(Perplexity(),))

Training CausalTransformer from step 0 to 300: 134.5M parameters, on 1 × NVIDIA A100-SXM4-40GB, batch 8, float32 parameters, bfloat16 compute

step  50/300  loss 3.000  ce 3.000  perplexity 20.08  token_accuracy 43.8%  step_time_ms 36.66  samples_per_sec 218.2  mfu 15.4%  accepted 100.0%  0:00:05 left

step 100/300  loss 3.165  ce 3.165  perplexity 23.69  token_accuracy 40.1%  step_time_ms 23.36  samples_per_sec 342.4  mfu 24.2%  accepted 100.0%  0:00:04 left

eval val at step 100: perplexity 30.04   loss 3.403 (14 records in 15.48 s)

step 150/300  loss 3.106  ce 3.106  perplexity 22.32  token_accuracy 43.0%  step_time_ms 384.2  samples_per_sec 20.82  mfu 1.5%  accepted 100.0%  0:00:22 left

step 200/300  loss 3.065  ce 3.065  perplexity 21.43  token_accuracy 43.1%  step_time_ms 23.41  samples_per_sec 341.7  mfu 24.1%  accepted 100.0%  0:00:12 left

eval val at step 200: perplexity 28.64 ↓ 4.7%   loss 3.355 ↓ 1.4% (14 records in 0.05 s)

step 250/300  loss 2.991  ce 2.991  perplexity 19.91  token_accuracy 43.0%  step_time_ms 98.69  samples_per_sec 81.07  mfu 5.7%  accepted 100.0%  0:00:06 left

step 300/300  loss 2.883  ce 2.883  perplexity 17.86  token_accuracy 42.8%  step_time_ms 23.37  samples_per_sec 342.3  mfu 24.2%  accepted 100.0%

eval val at step 300: perplexity 28.20 ↓ 1.5%   loss 3.339 ↓ 0.5% (14 records in 0.06 s)

Trained 300 steps in 0:01:02: first step after 24.73 s, then 20.7 step/s

23.2% of the wall time in steps, final loss 2.883

In [9]:
after = generate(model, state.variables, prompt, max_new_tokens=MAX_NEW_TOKENS,
                 key=jax.random.key(0), sampling=Sampling(temperature=0.0))
print(tokenizer.decode(after.tokens[0]))

ROMEO:
I am not so: I am not so.

FRIAR JOHN:
I am not so: I am not so.

FRIAR PETER:
I am not so: I am not so.

FRIAR MARCIUS:



Validation perplexity went from about 29 at step 100 to about 28 at step 300, a small change from so few steps. The greedy text changed a lot more though: the speakers are now characters from the plays, such as Duke Vincentio, and the lines have the play's layout. It still repeats itself, as greedy decoding tends to do.

## Exporting back to Hugging Face

`PretrainedDecoder.from_model` wraps the trained weights for export, and its `save` writes them in the Hugging Face layout: `config.json`, `model.safetensors`, and the tokenizer files. It runs the same name and shape translation as `PretrainedDecoder.load`, just backwards. Loading the export again gives back exactly the weights we trained.

In [10]:
PretrainedDecoder.from_model(model, state.variables, tokenizer=CHECKPOINT).save(EXPORT_DIR)
print(sorted(os.listdir(EXPORT_DIR)))

reloaded = PretrainedDecoder.load(EXPORT_DIR, dtype=jnp.bfloat16, attention_impl="auto", max_seq_len=512)
same = all(np.array_equal(np.asarray(a), np.asarray(b))
           for a, b in zip(jax.tree_util.tree_leaves(reloaded.variables["params"]),
                           jax.tree_util.tree_leaves(state.variables["params"])))
print("reloaded weights equal the trained ones:", same)

['config.json', 'generation_config.json', 'model.safetensors', 'tokenizer.json', 'tokenizer_config.json']


reloaded weights equal the trained ones: True


## Reading the export with `transformers`

An export is only useful if another library can actually read it. So we load the folder with `transformers`, run the prompt through it in float32 on the CPU, and compare its most likely next token with Dew's.

In [11]:
import torch
from transformers import AutoModelForCausalLM

hf_model = AutoModelForCausalLM.from_pretrained(EXPORT_DIR, dtype=torch.float32)
with torch.no_grad():
    hf_next = int(hf_model(input_ids=torch.tensor(np.asarray(prompt))).logits[0, -1].argmax())

dew_next = generate(model, state.variables, prompt, max_new_tokens=1,
                    key=jax.random.key(0), sampling=Sampling(temperature=0.0))
dew_next = int(dew_next.tokens[0, -1])
print("transformers:", hf_next, repr(tokenizer.decode([hf_next])))
print("dew:         ", dew_next, repr(tokenizer.decode([dew_next])))

Loading weights:   0%|          | 0/272 [00:00<?, ?it/s]

Loading weights: 100%|██████████| 272/272 [00:00<00:00, 5147.64it/s]

transformers: 198 '\n'
dew:          198 '\n'


## Where to go next

The [README model list](../README.md#models) names the decoder families Dew loads, trains, generates with and exports. The recipe flag `--pretrained` runs this notebook's flow at scale, for example `python recipes/lm/train.py data:token-windows --data.path <dir> --pretrained HuggingFaceTB/SmolLM2-135M --tokenizer HuggingFaceTB/SmolLM2-135M`.